File: src/p315/Solution.java
package p315;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

class SegmentTree {
	int nums[];
	int n;
	int tree[];
	int count[];
	int left[];
	int right[];

	public SegmentTree(int nums[]) {
		this.nums = nums;
		this.n = nums.length;
		buildTree();
	}

	public void buildTree() {
		int treeSize = (n << 2) + 4;
		tree = new int[treeSize];
		count = new int[treeSize];
		left = new int[treeSize];
		right = new int[treeSize];
		buildTree(0, 0, n - 1);
	}

	private void buildTree(int node, int left, int right) {
		this.left[node] = left;
		this.right[node] = right;
		if (left != right) {
			int mid = ((left + right) >> 1);
			buildTree(node + node + 1, left, mid);
			buildTree(node + node + 2, mid + 1, right);
		}
	}

	public void addNumber(int num) {
		addNumber(0, num);
	}

	private void addNumber(int node, int num) {
		count[node]++;
		if (left[node] == right[node])
			return;
		if (num <= nums[(left[node] + right[node]) >> 1])
			addNumber(node + node + 1, num);
		else
			addNumber(node + node + 2, num);
	}

	public int query(int num) {
		return query(0, num);
	}

	private int query(int node, int num) {
		if (nums[right[node]] < num)
			return count[node];
		if (nums[left[node]] >= num)
			return 0;
		return query(node + node + 1, num) + query(node + node + 2, num);
	}
}

public class Solution {
	public List<Integer> countSmaller(int[] nums) {
		if (nums.length == 0) {
			return Collections.emptyList();
		}
		int sortedNums[] = Arrays.copyOf(nums, nums.length);
		Arrays.sort(sortedNums);
		SegmentTree tree = new SegmentTree(sortedNums);
		List<Integer> list = new ArrayList<Integer>();
		for (int i = nums.length - 1; i >= 0; i--) {
			list.add(tree.query(nums[i]));
			tree.addNumber(nums[i]);
		}
		Collections.reverse(list);
		return list;
	}

	public static void main(String[] args) {
		int nums[] = { 1, 1 };
		for (int t : new Solution().countSmaller(nums)) {
			System.out.println(t);
		}
	}

}

File: src/p315/solution.py
class Solution(object):
    def merge(self, nums, left, mid, right):
        i, j = left, mid + 1
        l = []
        while i <= mid and j <= right:
            if nums[i][1] <= nums[j][1]:
                l.append(nums[i])
                self.result[nums[i][0]] += j - mid - 1
                i += 1
            else:
                l.append(nums[j])
                j += 1
        while i <= mid:
            l.append(nums[i])
            self.result[nums[i][0]] += right - mid
            i += 1
        while j <= right:
            l.append(nums[j])
            j += 1
        assert len(l) == right - left + 1
        for i in range(len(l)):
            nums[i + left] = l[i]

    def mergesort(self, nums, left, right):
        if left >= right:
            return 0
        mid = (left + right) >> 1
        self.mergesort(nums, left, mid)
        self.mergesort(nums, mid + 1, right)
        self.merge(nums, left, mid, right)

    def countSmaller(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        self.result = [0] * len(nums)
        self.mergesort(list(enumerate(nums)), 0, len(nums) - 1)
        return self.result


